File: LeetCode/cpp/264.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        vector<int> primes{2, 3, 5};
        int m = primes.size();
        vector<int> factor(primes.size(), 0), res(n, 1);
        for (int i = 1; i < n; ++i) {
            res[i] = INT_MAX;
            for (int j = 0; j < m; ++j)
                res[i] = min(res[i], primes[j] * res[factor[j]]);
            for (int j = 0; j < m; ++j)
                if (primes[j] * res[factor[j]] == res[i])
                    ++factor[j];
        }
        return res.back();
    }
};

File: LeetCode/python/264.py
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        primes = [2, 3, 5]
        m = len(primes)
        factor, res = [0 for _ in range(m)], [1 for _ in range(n)]
        for i in range(1, n):
            res[i] = sys.maxsize
            for j in range(m):
                res[i] = min(res[i], primes[j] * res[factor[j]])
            for j in range(m):
                if primes[j] * res[factor[j]] == res[i]:
                    factor[j] += 1
        return res[-1]
